### ISSUE TITLE
Q: Can't understand LogLevel settings.

### ISSUE THREAD

**Issue by bernatgy** [adams85/filelogger#11]
Hi!

This is just a question, or maybe an enhancement for the docs if it's not just me.

First of all, thank you for this library! Good job! I've been using it for about 6 months, had no problem!

Context:
**Library Version:** 3.0.0 _(I'll probably update soon...)_
**Target framework:** .NET Core 3.0 _(Upgrade coming as well...)_
_It's a complex ASP.NET Core project, with background tasks etc._

So the one thing I've been messing around is setting up my logs to be "just perfect".
I have some **Global LogLevel settings**:
```JSON
"Logging": {
    ...
    "LogLevel": {
      "Default": "Trace",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information",
      "NHibernate": "Debug"
    },
    ...
}
```

Then I have a Console provider which is most probably irrelevant.
And lastly I have my **File provider settings**:
```JSON
"File": {
      "BasePath": "logs",
      "FileAccessMode": "KeepOpenAndAutoFlush",
      "MaxFileSize": 10485760,
      "CounterFormat": "000",
      "IncludeScopes": true,
      "MaxQueueSize": 1000,
      "Files": [
        {
          "Path": "app-persistence-<counter>.log",
          "MinLevel": {
            "Default": "Trace",
            "NHibernate": "Debug",
            "NHibernate.Type": "Information",
            "NHibernate.Cfg.XmlHbmBinding": "Warning",
            "NHibernate.Cfg.XmlHbmBinding.Binder": "Information",
            "Itg.Persistence": "Trace"
          }
        },
        {
          "Path": "app-<counter>.log",
          "MinLevel": {
            "Default": "Debug",
            "Microsoft": "Warning",
            "Microsoft.AspNetCore.Authentication": "Debug",
            "Microsoft.AspNetCore.Authorization": "Debug",
            "Microsoft.AspNetCore.Server": "Debug",
            "NHibernate": "None",
            "Itg.Persistence": "None",
            "Itg.Services": "Debug",
            "Itg.Services.Shop": "Information",
            "Itg.Mailing": "Information"
          }
        }
      ]
    }
```

The thing I can't seem to understand is the following:
I'd like my _app-persistence-\<counter\>.log_ to contain only `NHibernate` and `Itg.Persistence` categories, the other one mostly everything else.

On my initial run I had `"Default": "None"` on my persistence file, which I thought must be right, because I didn't want anything else to land in there, only the ones I specified (*and more specific ones!*).

Now based on these, after reading through Microsoft's logging tutorial, and this repo's readme, the category `"Itg.Persistence.Secondary"` should have been written to the persistence file, but it did not! (it's "under" `"Itg.Persistence"` which I set to `"Trace"`)

The logging worked, as I could find my entries on the console, but nowhere else.
To be exact, this is the line I was looking for:
`crit: Itg.Persistence.Secondary.FileKeyValueStore[LoadFailed]`
And to be more exact, this line is logged by a queued background task!

Now I set `"Default": "Trace"` on my persistence file, and here I thought, now everything should show up. Not just the ones I specified. To my surprise the one I was looking for *DID* show up (`"Itg.Persistence.Secondary"`) *BUT* not at all everything... This is the only new category in my persistence log. Which is just about the effect I wanted, but I can't imagine these settings are any good...

Could you help me out please? What am I missing? I'm so confused... :(

**Comment by bernatgy** [comment]
I must have forgot to delete the log at one point.
**Correction;**
With the whole configuration I copied in, still nothing is showing up in my persistence log, other than `"NHibernate*"` categories.

**Comment by adams85** [comment]
Hi!

First of all, thanks for the kind words. :)

I think your understanding of log level settings are fundamentally correct. There's one important thing to keep in mind. Log level switches defined at log file level (`MinLevel` option) can only make the log level switches defined at higher levels (global level or log provider level) stricter. This is mentioned in the docs as well:

> [...] this is a second-level filter which can only tighten the rules of the first-level LogLevel filters (as it can only filter messages that hit the logger provider).

This is a limitation of the _Microsoft.Extensions.Logging_ framework I cannot do about anything as it's out of my control. However, in your case this cannot cause any problems at all (theoretically).

That is, your configuration should work. I put together a [simplified imitation of your environment](https://gist.github.com/adams85/287de6283948d6f2f63e0b9cf3025fb7) based on your description but I failed to reproduce the issue. It just works as expected.

Please take a look at my attempt and try to figure out if there's some relevant differences between my simulated and your real environment which I don't see.

**Comment by bernatgy** [comment]
Hi! I will take a look, I promise! Thank you! Just letting you know I saw this. I will edit this post before monday (25. May)!

**Comment by adams85** [comment]
I think the key momentum here is the blockage of the Itg.Services.* messages. However, I cannot see yet how they can get stuck for minutes. I suspect though it has something to do with the relatively high logging traffic of NHibernate.

Could you do a test where you fire a single(!) request and check the logs after some minutes if the Itg.Services.* entries are still missing? I'd just like to make sure that we're dealing with just one or two issues here (that is, whether there's also something regarding configuration reloading or not).

If the test failed (no Itg.Services.* entries after say 5 minutes), then please try the following:
1. Repeat the test described above with unbounded queues (`MaxQueueSize` set to 0)
2. Repeat the test where `FileAccessMode` set to `KeepOpen` and `OpenTemporarily`. Let's see if this makes a difference. (There's [an issue with flushing file streams asynchronously](https://github.com/dotnet/runtime/issues/27643) in .NET. I ran into perf problems because of this but it's been solved - theoretically.)
3. Update the file logger lib to the latest version (3.1.1). You can do this safely, I keep to semver (no breaking changes to the public APIs in minor versions). 3.1 comes with some perf-related improvements (e.g. [Dataflow was replaced by Channels](https://github.com/adams85/filelogger/issues/5)). Let's see if this changes anything.

This will take some time of yours, but I'm afraid we cannot do anything else. We have to narrow down the possible causes of your issue to be able to track it down.

**Comment by adams85** [comment]
I updated [my repro app](https://gist.github.com/adams85/287de6283948d6f2f63e0b9cf3025fb7) according to the new details you supplied. It feeds 10k log entries into the logger and does some occasional logging in the background.

All entries land in the log as expected and I experienced no blocking. (`KeepOpen` vs. `KeepOpenAndAutoFlush` makes a significant difference though!)

**Comment by bernatgy** [comment]
Hi, and thank you again!

_Interesting 1.:_ The non-persistence file seems to always work correctly, though I did not edit it's settings.
_Interesting 2.:_ I cannot set `MaxQueueSize` to 0. An exception gets thrown from filelogger on the first log event. I could only remove the setting alltogether (which should set it to 0, but this is confusing...). I can tell you more about it, but I'm guessing the result will be to upgrade to the new version either way, which would probably invalidate it, so let's wait it out. 😊

I am of course doing the same (single!) request in every case, with same parameters, so logs should be - and are - the same, and I've also been doing the waiting thing. 🤔 I wait on average 10~15 minutes for the logs, they just never come.

**A bit of a confusion-correction:**
I want `Itg.Persistence.Secondary` logs in my Persistence file (not `Itg.Services.*`) :)
At first, when the file reloaded, both landed in there, I'm assuming because of the `"Default": "Trace"` setting.
But later on, I set `"Default": "None"`, and now `Itg.Services` does not appear in the Persistence file, not initially and not on reload, which is what I want.
_(those ones go into the other log file, correctly)_

Right now, the correct behaviour is happening, which is that NHibernate lines go into the Persistence file, and `Itg.Persistence.Secondary` lines go there as well, **BUT** only on reload (delay).

**Note:**
Throughout this thing, I set `FileKeyValueStore` to be a singleton. (actually, your repro reminded me, that's more fitting) _This had no effect on the observed behaviour._

**Here are the cases you requested:**
- `KeepOpenAndAutoFlush` and removed `MaxQueueSize`: No change in behaviour.  _(initially, or on reload)_
- `KeepOpen` and removed `MaxQueueSize`: No change in behaviour. _(initially, or on reload)_
- `KeepOpen` and `MaxQueueSize: 10000` : No change in behaviour. _(initially, or on reload)_
- `OpenTemporarily` and removed `MaxQueueSize`: **THIS SOLVED IT!!!!! (I'm writing this in real time so... 😂 checking again... yep! solved!)**
- `OpenTemporarily` and `MaxQueueSize: 10000` : **STILL WORKS!**

This feels so weird. Why the `FileKeyValueStore`...? Could it be that, it's also manipulating a different file? Could that specific thread be blocked by my file manipulations?
And I mean blocked this hard? It's horrible code... But works... I can share some if you'd like...

I'm going to wait for some kind of a response, and am going to update to `3.1.1`!!!
Thank you very much for your continued support!

**Comment by adams85** [comment]
> I cannot set MaxQueueSize to 0. An exception gets thrown from filelogger on the first log event.

It's my bad. Using v3.0 you have to specify a negative number (like -1) to get unbounded queues (or just remove that setting altogether). v3.1 changed this: 0 is also treated as unbounded.

> This feels so weird. Why the FileKeyValueStore...? Could it be that, it's also manipulating a different file? Could that specific thread be blocked by my file manipulations?

I think we're getting somewhere. `OpenTemporarily` opens, writes and closes the file for each log entry. (Needless to say, how this affects performance. It should only be used when low logging traffic is expected.) In contrast to this, `KeepOpen` and `KeepOpenAndAutoFlush` opens a stream to the log file once and uses that for writing all the incoming log entries. **Writes are done on thread pool threads using async I/O.** Except for flush (as I mentioned earlier) which needs to be a synchronous call for the moment. I thought there's something with this but you wrote that `KeepOpen` doesn't work either. That mode involves no flushing. The file stream is created with [these settings](https://github.com/adams85/filelogger/blob/3.1.1/source/FileLogger/FileAppender.cs#L55). Speaking of such, what OS do you run your tests on?

I don't see yet how your code could interfere with the log writer but it seems this is the case. Is there any chance that your app opens the log files being written?

BTW, please upgrade to 3.1.1, repeat the tests and let's see if the issue persists. In the meantime, I'll also do a few tests by adding some file access simulation to the repro app.

**Comment by adams85** [comment]
Just a thought flashed through my mind. Do you run some anti-virus software? If so, do a test with the AV software disabled. There's not much chance that it cause an interference but let's eliminate that for sure.

**Comment by bernatgy** [comment]
Hi!
Again with the delay... 😢 A partial response for now:

Here's my system:
![image](https://user-images.githubusercontent.com/12005276/83273001-3b638680-a1cc-11ea-8378-f5e3e5845885.png)

But actually, I've had a multiple build update through an in-place reinstall since opening the issue, and it still did not change!

Also - while all this is still happening by chance which is **crazy** - at first I didn't have any antivirus! I had Windows Defender, and it just went dead and unnoticed. I couldn't start it because of some Windows Update corruption mess... That's why I then reinstalled Windows, and I'm now proudly using Malwarebytes. So, I've actually tried this with almost all possibilities in terms of an antivirus and OS version! 😂😂

If you're thinking what if all this madness played a part... My OS drive is not my work drive (which I have this project on). Nothing really happed on the work drive...

> OpenTemporarily opens, writes and closes the file for each log entry. (Needless to say, how this affects performance. It should only be used when low logging traffic is expected.)

That's not good... 😢
I'll get back to you probably tomorrow after the upgrade!

**Comment by bernatgy** [comment]
Upgraded to `3.1.1` and did the tests again _(with Malwarebytes and Defender turned off)_.
Nothing seems to have changed. 🤔

`OpenTemporarily` still works.
`KeepOpen` and `KeepOpenAndAutoFlush` still doesn't, but they spit the logs I want when the config reloads.
_(Checked with 10k MaxQueueSize and without one)_

[This is my whole `FileKeyValueStore` implementation](https://gist.github.com/bernatgy/f4abf64d6ab918d0bc2ecec4da272652) (reminder, excuse the mess! 😂)
I've commented on the lines I'd expect to see in my log.

About _2kb_ of binary serialized data into store file per request.
It's 686kb at the moment.
Saving takes about a second at this point. _(judging by entries on console)_

Now deleting the store file and doing it again.
No difference.

**Comment by adams85** [comment]
Sorry for the delay, I've just found time for continuing the investigation.

I created [a much more sophisticated repro app](https://github.com/adams85/filelogger/tree/repro/issue11/test/issue11). I used the code snippets you provided and did my best to fill in the gaps and reconstruct your context as close as possible.

This was the good news. The bad news is that I still cannot reproduce the problem... It works as expected. So please examine my new attempt and try to identify the differences between that and your real-life app.

After starting the repro app, you can simulate a request by visiting http://localhost:5000/Test. Please debug through the application and check for relevant differences. The entry point is `MainController.Test`. This action will enqueue a work item which is processed by the background worker (`QueuedHostedService`). From that point on the code will certainly be very familiar to you.

Let's see if this brings us a result.

**Comment by bernatgy** [comment]
Hi! I'm so sorry for the delay on my part as well.

So I've been trying to reproduce the issue in the repro app you created. I've gone ahead and made it closer and closer to my actual flow and implementation, but I've kind of given up. The only thing I did not modify yet is the setup in the Program class. Mine is a lot different because I'm not using the default builder. (Host.CreateDefaultBuilder) I'm doing it manually. Will check this out, but I think I've found a lot better lead just now.

I've removed the filelogger NuGet package reference, went ahead and added the source project as dependency! Currently I'm debugging through filelogger, and I've already identified the symptom.

The log file is successfully identified and everything actually goes smooth through [`FileLoggerProcessor.Enqueue(...)`](https://github.com/adams85/filelogger/blob/2af30b124b4cbf0fd16cd9ce826d5a13d0f81a81/source/FileLogger/FileLoggerProcessor.cs#L234) but then the FileLogEntry is just stuck in the [Queue](https://github.com/adams85/filelogger/blob/2af30b124b4cbf0fd16cd9ce826d5a13d0f81a81/source/FileLogger/FileLoggerProcessor.cs#L54) forever. 

If I just let the debug through and perform another request, I can actually see that all the previous entries are still in the Queue.
![image](https://user-images.githubusercontent.com/12005276/85204032-f2f24100-b311-11ea-9e4b-56b0a1d47a96.png)

Then I saw that every exception is swallowed in [`FileLoggerProcessor.WriteEntryAsync`](https://github.com/adams85/filelogger/blob/2af30b124b4cbf0fd16cd9ce826d5a13d0f81a81/source/FileLogger/FileLoggerProcessor.cs#L347) which put my mental state in a weird place lol. Just dropped a breakpoint on the catch blocks and as foretold by great oracles of the past; there's always an exception...

`IOException: The process cannot access the file 'D:\\...\\integra-persistence-000.log' because it is being used by another process.`
_at FileLoggerProcessor:382_

I've put up a static InstancesCount field which I'm incrementing in the constructor, and decrementing at the end of `Dispose()`, because I'm sure I don't have the file open in anything.
Ran through debugging, there's two instances of `FileLoggerProcessor` when the exception is thrown.

At this point I don't really have a clue how that could happen. Logging is registered as Singleton for a good reason.

Do you think the following line in my `QueuedHostedService` (_which I've shared [here](https://github.com/adams85/filelogger/issues/11#issuecomment-630774960)_) could somehow create two completely different service providers?
```csharp
using var scope = this._services.CreateScope();
```

Seems pretty strange... Docs say;
> Creates a new IServiceScope that can be used to resolve scoped services.

Do you have an idea about where I could go deeper into this? I'm a complete newbie to debugging IO and Threading to be honest.

Also, pretty please... Could we have some kind of an indication about the exceptions? 😢😢😢😢
At least highlighted messages to the Debug out (while I realize this is utterly useless in production)? Or literally anything else? I know this is a really sensitive topic. What a logging solution should or should not do about / with exceptions is hard to determine but in my opinion, straight up swallowing all of them is not the best thing to do.

Again, I deeply appreciate your time and effort in this adventure!
Thanks in advance!

**Comment by adams85** [comment]
> I've removed the filelogger NuGet package reference, went ahead and added the source project as dependency! Currently I'm debugging through filelogger, and I've already identified the symptom.

After failing to reproduce the issue, this would have been my last resort as well. I'm glad that you've already done that. 👍

> Ran through debugging, there's two instances of FileLoggerProcessor when the exception is thrown.

This explains why `OpenTemporarily` works and `KeepOpen*` doesn't. In the latter case, both processors keep the log files open but only the "lucky" one which opens it *sooner* can actually write into it because of `FileShare.Read`. The "unlucky" one cannot write its queue to the file until the other one releases it, which explains why the missing log entries appear suddenly when you edit the settings (log files are always released on reloading of the settings).

Now the question is how on earth you end up with two `FileLoggerProcessor` instances. This is possible only if you have two instances of `FileLoggerProvider` as well. So I'm pretty sure that this is a DI configuration problem on your side.

* Does your application, by chance, use multiple DI containers?
* If not, could you verify the number of `FileLoggerProvider`s registered in your DI container? E.g. you can do this by adding the following snippet to your `Startup.Configure` method:
    ```csharp
    System.Diagnostics.Debug.WriteLine("# of file logger providers in root scope: " + app.ApplicationServices.GetRequiredService<IEnumerable<ILoggerProvider>>().OfType<FileLoggerProvider>().Count());
    using (var scope = app.ApplicationServices.CreateScope())
        System.Diagnostics.Debug.WriteLine("# of file logger providers in child scope: " + scope.ServiceProvider.GetRequiredService<IEnumerable<ILoggerProvider>>().OfType<FileLoggerProvider>().Count());
    ```

> Could we have some kind of an indication about the exceptions? 😢😢😢😢
At least highlighted messages to the Debug out (while I realize this is utterly useless in production)?

I'm [aware of this issue](https://github.com/adams85/filelogger/issues/8#issuecomment-614610284) but I haven't figured out yet how to solve it. The crux of it is that we want to do logging in a logger component. Calls to `Debug.WriteLine` are stripped out in release builds so that won't work. `Trace.WriteLine` may work though. But it might be even better to add some event by means of which one can observe the swallowed exceptions if necessary.

**Comment by bernatgy** [comment]
Hi!

I'm not using multiple DI containers.
Your guess was correct though, I have 2 instances of `FileLoggerProvider` as well.
I'm currently tracking down how the second one could be created... A few hours in, but no luck yet.
The only thing I found out, was that using your snippet in `Startup.Configure`, both lines report only 1 instance.

At the same time, I have a breakpoint in the constructor of `FileLoggerProcessor` and I clearly see a second instance created _by the DI, guessing from the call stack_.

The first one is created by the `AddFile(...)` call in `.ConfigureLogging(...)`. I could not yet identify where the second one comes from, but the funny thing is that, the ones found by your snippet already have these second instances.

(I put the instance index as a field onto them...)
Will update once I figure something out. Thanks!

**Comment by adams85** [comment]
> I have an ILoggerFactory injected into my Startup class's constructor.

I don't think this is a good idea, at least, it's nonstandard. The framework allows injecting [these services](https://github.com/dotnet/aspnetcore/blob/v3.1.5/src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs#L355) in `Startup` currently. (Not sure about older versions though.)

> 
> I'm 99% percent sure this has to do with the fact that I'm still using the old .NET Core 2.x style WebHostBuilder setup instead of the new CreateDefaultBuilder / ConfigureWebHostDefaults and the relation between these and the internal IStartup interface.
> 
> At the time I made the conscious decision to stick with the old one, as I went through the source of the new methods and thought I could just pretty much copy / paste most of what they do.

I suspect that during your customized host setup an intermediate DI container is built somewhere. When you inject `ILoggerFactory` in your `Startup` ctor, it gets resolved by this intermediate container, hence an unwanted `IFileLoggerProvider` instance gets created.

I recommend dropping the custom solution and migrating to the ASP.NET Core 3-style (generic host) setup. AFAIK, logging should be configured through `ILoggingBuilder`, configuring `ILoggerFactory` directly has been discouraged for a while now. However, if you absolutely need it, in theory, you can do something like this:

```csharp
public static void Main(string[] args)
{
    var host = CreateHostBuilder(args).Build();
    ConfigureLoggerFactory(host.Services.GetRequiredService<ILoggerFactory>());
    host.Run();
}

private static void ConfigureLoggerFactory(ILoggerFactory loggerFactory)
{
    loggerFactory.AddProvider(...);
}
```
